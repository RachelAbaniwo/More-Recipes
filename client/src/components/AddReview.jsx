import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { createReview } from '../store/actions/recipes';


/**
 * @class AddReview
 * @extends {React.Component}
 * @description renders the add-review component
 *
 * @returns {textarea} review textarea field and button
 */
export class AddReview extends React.Component {
  /**
   * Adds review
   * @constructor
   *
   * @param {object} props
   *
   * @returns {void} void
   */
  constructor(props) {
    super(props);
    this.state = {
      review: ''
    };

    this.handleChange = this.handleChange.bind(this);
    this.createReview = this.createReview.bind(this);
  }
  /**
   * Creates review and sets state with empty review textarea field
   * @method createReview
   *
   * @param {null} null
   *
   * @returns {void} void
   */
  createReview() {
    this.props.createReview(this.state.review, this.props.recipeId);
    this.setState({
      review: ''
    });
  }
  /**
   * Handles change in review field, listens for change event in the review
   * textarea
   * @method handleChange
   *
   * @param {object} event
   *
   * @returns {void} void
   */
  handleChange(event) {
    this.setState({
      review: event.target.value
    });
  }
  /**
   * Renders add review jsx
   * @memberof AddReview
   *
   * @returns {jsx} jsx
   */
  render() {
    return (
      <div className="form-group">
        <textarea
          cols="120"
          onChange={this.handleChange}
          className="form-control mb-3"
          value={this.state.review}
          placeholder="Add review ..."
        />
        <button
          className="review btn btn-default title"
          disabled={this.state.review.length < 8}
          onClick={this.createReview}
        >ADD REVIEW
        </button>
      </div>
    );
  }
}
AddReview.propTypes = {
  recipeId: PropTypes.number.isRequired,
  createReview: PropTypes.func.isRequired
};
/**
 * Map dispatch to props
 * @function  mapDispatchToProps
 * @param {object} dispatch
 *
 * @returns {object} object to be passed as props to component
 */
export const mapDispatchToProps = dispatch => bindActionCreators({ createReview }, dispatch);
export default connect(null, mapDispatchToProps)(AddReview);

