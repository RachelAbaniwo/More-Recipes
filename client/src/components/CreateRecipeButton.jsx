import React from 'react';
import { Link } from 'react-router';

/**
 * Landing page create recipe buttons component
 * @function
 *
 * @param {object} authUser
 * @returns {object} jsx
 *
 */
const CreateRecipeButton = ({ authUser }) => {
  if (authUser === null || authUser === undefined) {
    return (<div style={{ marginTop: 50 }} />);
  }
  if (authUser !== null && authUser !== undefined) {
    return (
      <div className="justify-content-center" style={{ marginTop: 30 }} >
        <Link
          to="/recipes/create"
          className="button btn btn-default link-button justify-content-center title"
          style={{
        width: 180, fontFamily: 'Advent Pro', fontSize: '20px'
      }}
        >CREATE A RECIPE
        </Link>
      </div>
    );
  }
};

export default CreateRecipeButton;
