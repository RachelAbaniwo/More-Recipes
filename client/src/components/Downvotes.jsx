import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { toggleVote } from './../store/actions/votes';
/**
 * Downvotes component
 * @param {object} downvotes
 *
 * @returns {object} jsx for component
 */
const Downvotes = ({ downvotes, toggleVote: toggleVoteFunc, recipeId }) => ((
  <Fragment>
    <i
      className="fa fa-thumbs-down text-danger"
      onClick={() => { toggleVoteFunc(recipeId, 'downvote'); }}
      aria-hidden="true"
    />{downvotes}&nbsp;
  </Fragment>
));
Downvotes.propTypes = {
  downvotes: PropTypes.number.isRequired,
  toggleVote: PropTypes.func.isRequired,
  recipeId: PropTypes.number.isRequired
};
/**
 * Map dispatch to component props
 * @param {function} dispatch
 *
 * @returns {object} object to be passed as props to component
 */
const mapDispatchToProps = dispatch => bindActionCreators({ toggleVote }, dispatch);

const DownvotesConnected = connect(null, mapDispatchToProps)(Downvotes);

export default DownvotesConnected;
