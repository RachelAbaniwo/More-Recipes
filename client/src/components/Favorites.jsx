import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { toggleFavorite } from './../store/actions/votes';

/**
 * Add favorites component
 * @function Favorites
 *
 * @param {object} props
 *
 * @returns {object} jsx for component
 */
export const Favorites = ({
  toggleFavorite: toggleFavoritesFunc, recipeId, authUser, hasFavorited
}) => ((
  <Fragment>
    <i
      className={`fa fa-heart click-add ${authUser && hasFavorited ? 'orange' : 'grey'}`}
      onClick={() => { toggleFavoritesFunc(recipeId); }}
      aria-hidden="true"
    />&nbsp;
  </Fragment>
));

Favorites.propTypes = {
  toggleFavorite: PropTypes.func.isRequired,
  recipeId: PropTypes.number.isRequired,
  hasFavorited: PropTypes.bool.isRequired,
  authUser: PropTypes.shape({
    user: PropTypes.shape({
      id: PropTypes.number.isRequired,
      username: PropTypes.string.isRequired,
      email: PropTypes.string.isRequired,
      aboutMe: PropTypes.string.isRequired
    })
  }),
};
Favorites.defaultProps = {
  authUser: null
};

/**
 * Map dispatch to component props
 * @param {function} dispatch
 *
 * @returns {object} object to be passed as props to component
 */
export const mapDispatchToProps = dispatch =>
  bindActionCreators({ toggleFavorite }, dispatch);

const FavoritesConnected = connect(null, mapDispatchToProps)(Favorites);

export default FavoritesConnected;
