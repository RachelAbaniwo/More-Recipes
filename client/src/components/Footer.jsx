
import React from 'react';
/**
 * Display footer component
 * @param {null} null
 *
 * @returns {object} jsx for footer
 *
 */
const Footer = () => ((
  <footer
    className="fixed-bottom"
    style={{
      backgroundColor: 'rgba(73, 67, 67, 0.9)',
      maxWidth: '100%',
      height: 30
    }}
  >
    <p
      className="text-center"
      style={{
          marginBottom: 5,
          paddingTop: 5,
          color: '#fff',
          fontSize: 12
        }}
    >
          Â© 2017 More-Recipes
    </p>
  </footer>
));

export default Footer;
