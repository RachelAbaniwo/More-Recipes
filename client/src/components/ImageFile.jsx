import React from 'react';
import PropTypes from 'prop-types';
import Dropzone from 'react-dropzone';

/**
 * Add a recipe image component
 * @class
 *
 * @returns {object} jsx object
 */
export class ImageFile extends React.Component {
  /**
   * Adds recipe image
   * @constructor
   *
   * @param {object} props
   *
   * @returns {object} jsx
   */
  constructor(props) {
    super(props);

    this.handleDrop = this.handleDrop.bind(this);
    this.buildImageMessage = this.buildImageMessage.bind(this);
  }

  /**
   * sets image url
   * @function
   *
   * @param {object} files
   *
   * @returns {function} file[0] as parameter
  */
  handleDrop(files) {
    this.props.setImageUrl(files[0]);
  }

  /**
   * Displays appropriate message alongside image
   * @function
   *
   * @param {null} null
   *
   * @returns {jsx} imgMsg jsx that displays message
  */
  buildImageMessage() {
    let imgMsg = null;
    if (!this.props.imageFile && !this.props.imageUrl) {
      imgMsg = (
        <div
          className="justify-content-center"
          style={{ height: '100px' }}
        >
          <h5
            className="justify-content-center"
          >Click to select Image
          </h5>
        </div>
      );
    } else {
      imgMsg = (
        <div className="justify-content-center">
          <div className="hovereffect">
            <img
              className="img-responsive col-sm-12 justify-content-center figure-img img-thumbnail image-fluid mx-auto"
              style={{ maxWidth: 700, maxHeight: 500 }}
              src={this.props.imageFile ? this.props.imageFile.preview : this.props.imageUrl}
              alt=""
            />
            <div className="overlay">
              <h2>Click to select another image</h2>
            </div>
          </div>
        </div>
      );
    }
    return imgMsg;
  }

  /**
   * Displays preview of recipe image
   *
   * @returns {jsx} jsx with dropzone mounting preview of image
   */
  render() {
    return (
      <Dropzone onDrop={this.handleDrop} style={{ width: '100%' }}>
        {this.buildImageMessage()}
      </Dropzone>
    );
  }
}
ImageFile.propTypes = {
  setImageUrl: PropTypes.func.isRequired,
  imageFile: PropTypes.shape({
    preview: PropTypes.string.isRequired
  }),
  imageUrl: PropTypes.string
};

ImageFile.defaultProps = {
  imageFile: null,
  imageUrl: null
};

export default ImageFile;
