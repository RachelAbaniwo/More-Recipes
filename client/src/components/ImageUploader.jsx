import React from 'react';
import { connect } from 'react-redux';
import Dropzone from 'react-dropzone';
import { bindActionCreators } from 'redux';
import { setImageUrl } from '../store/actions';
import foodPlaceholder from '../../assets/image/food-placeholder.jpeg';

import '../../assets/css/hover.css';

class ImageFile extends React.Component {
  constructor(props) {
    super(props);

    this.handleDrop = this.handleDrop.bind(this);
    this.buildImageMessage = this.buildImageMessage.bind(this);
  }
  handleDrop(files) {
    this.props.setImageUrl(files[0]);
  }
  buildImageMessage() {
    let imgMsg = null;
    if (!this.props.imageFile) {
      imgMsg = (
        <div className="justify-content-center" style={{ height: '100px' }}>
          <h5 className="justify-content-center">Click to select your Recipe Image</h5>
      </div>
      );
    } else {
      imgMsg = (
          <div className="justify-content-center">
            <div className="hovereffect">
              <img className="img-responsive col-sm-12 justify-content-center figure-img img-thumbnail image-fluid mx-auto" 
                    style={{ maxWidth: 700, maxHeight: 500 }}
                   src={this.props.imageFile.preview} alt="" />
              <div className="overlay">
                <h2>Click to upload another image</h2>
              </div>
          </div>
        </div>
      );
    }
    return imgMsg;
  }

  render() {
    return (
      <Dropzone onDrop={this.handleDrop} style={{ width: '100%' }}>
        {this.buildImageMessage()}
      </Dropzone>
    );
  }
}



const mapStateToProps = state => {
  return {
    imageFile: state.imageUpload.imageFile
  };
};


const mapDispatchToProps = dispatch => {
  return bindActionCreators(
    { setImageUrl }, dispatch
  );
};

const ImageUrlContainer = connect(mapStateToProps, mapDispatchToProps)(ImageFile);

export default ImageUrlContainer;
