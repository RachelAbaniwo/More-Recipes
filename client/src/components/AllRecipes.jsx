import React from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { Link } from 'react-router';

import Footer from './Footer';
import { getAllRecipes } from '../store/actions/recipes';

import SingleRecipe from './SingleRecipe';

/**
 * Displays all recipes
 * @class
 *
 * @returns {object} jsx object
 */
class AllRecipeScreen extends React.Component {
  /**
   * Execute before component mounts
   * @returns {null} null
   */
  componentWillMount() {
    this.props.getAllRecipes();
  }
  /**
   * Display recipes
   *
   * @returns {jsx} jsx
   */
  render() {
    const { recipes } = this.props;
    return (
      <div>
        <section id="nav">
          <nav
            className="navbar navbar-expand-sm navbar-dark fixed-top navbar-custom"
          >
            <Link
              to="/home"
              className="moreRecipes"
            > MORE RECIPES
            </Link>
            <button
              className="navbar-toggler"
              type="button"
              data-toggle="collapse"
              data-target="#navbarSupportedContent"
              aria-controls="navbarSupportedContent"
              aria-expanded="false"
              aria-label="Toggle navigation"
            >
              <span className="navbar-toggler-icon" />
            </button>
            <div
              className="collapse navbar-collapse"
              id="navbarSupportedContent"
            >
              <ul className="navbar-nav ml-auto">
                <li className="nav-item">
                  <Link
                    to="/signup"
                    className="navbar-register"
                    style={{ marginRight: 20 }}
                  >REGISTER
                  </Link>
                </li>
                <li className="nav-item">
                  <Link
                    to="/signin"
                    className="navbar-register"
                    style={{ marginRight: 20 }}
                  >SIGN IN
                  </Link>
                </li>
              </ul>
              <form className="form-inline my-2 my-lg-0">
                <input
                  className="form-control mr-sm-2"
                  type="search"
                  placeholder="Find a Recipe"
                  aria-label="Search"
                />
              </form>
            </div>
          </nav>
        </section>
        <section
          className="container text-center mx auto view-recipe-container"
          style={{
            border: '1px solid lightgrey',
            padding: 30,
            marginTop: 90,
            marginBottom: 50
          }}
        >
          <div className="row">
            <section className="col-sm-12">
              <h1>
                All Recipes
              </h1>
            </section>
            <div className="row">
              <form className="form-inline col-lg-12 justify-content-center">
                <input
                  className="form-control ml-sm-4 mr-sm-4"
                  style={{ width: 250, fontFamily: '"kaushan script"' }}
                  type="search"
                  placeholder="Search Recipes"
                  aria-label="Search"
                />
                <select
                  className="custom-select mb-2 ml-sm-4 mr-sm-4 mb-sm-0"
                  style={{ width: 250, fontFamily: '"kaushan script"' }}
                  id="inlineFormCustomSelect"
                >
                  <option selected>Sort By...</option>
                  <option value={1}>Most Upvotes</option>
                  <option value={2}>Most Favorited</option>
                </select>
                <select className="custom-select mb-2 ml-sm-4 mr-sm-4 mb-sm-0" style={{ width: 250, fontFamily: '"kaushan script"' }} id="inlineFormCustomSelect">
                  <option selected>Recipes per page</option>
                  <option value={1}>3</option>
                  <option value={2}>6</option>
                  <option value={3}>9</option>
                </select>
                <button
                  style={{ width: 150 }}
                  type="submit"
                  className="btn btn-default"
                >Search
                </button>
              </form>
            </div>


            <div className="row" style={{ padding: '10px' }}>
              {recipes.map(recipe => <SingleRecipe key={recipe.id} recipe={recipe} />)}
            </div>
          </div>
        </section>
        <Footer />

      </div>
    );
  }
}

AllRecipeScreen.propTypes = {
  recipes: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    ingredients: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    category: PropTypes.string.isRequired,
    method: PropTypes.string.isRequired,
    userId: PropTypes.number.isRequired,
    views: PropTypes.number.isRequired,
    favorites: PropTypes.number.isRequired,
    upvotes: PropTypes.number.isRequired,
    downvotes: PropTypes.number.isRequired,
    createdAt: PropTypes.string.isRequired,
    updatedAt: PropTypes.string.isRequired,
  })),
  getAllRecipes: PropTypes.func.isRequired
};

AllRecipeScreen.defaultProps = {
  recipes: []
};

/**
 * Map state to props
 * @param {object} state
 *
 * @returns {object} object of recipes passed as props
 */
const mapStateToProps = state => ({ recipes: state.recipes.rows });

/**
 * Map dispatch to props
 * @param {func} dispatch
 *
 * @returns {object} object to be passed as props to component
 */
const mapDispatchToProps = dispatch => bindActionCreators({ getAllRecipes }, dispatch);

const AllRecipeScreenContainer = connect(mapStateToProps, mapDispatchToProps)(AllRecipeScreen);

export default AllRecipeScreenContainer;
