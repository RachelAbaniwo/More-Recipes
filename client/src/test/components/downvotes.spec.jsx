import React from 'react';
import { mount } from 'enzyme';
import { Downvotes, mapDispatchToProps } from '../../components/Downvotes';

describe('The Downvote component', () => {
  const props = {
    downvotes: 1,
    recipeId: 1,
    toggleVote: jest.fn()
  };

  it('should mount without crashing', () => {
    const wrapper = mount(<Downvotes {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
  it('should toggle vote count if user clicks to downvote', () => {
    const wrapper = mount(<Downvotes {...props} />);
    wrapper.find('i').simulate('click');
    expect(props.toggleVote).toHaveBeenCalledWith(1, 'downvote');
  });
});

describe('The Downvote mapDispatchToProps', () => {
  it('should return toggleVote action', () => {
    const dispatch = jest.fn();

    const results = mapDispatchToProps(dispatch);
    expect(results.toggleVote).toBeTruthy();
    expect(typeof results.toggleVote).toBe('function');
  });
});
