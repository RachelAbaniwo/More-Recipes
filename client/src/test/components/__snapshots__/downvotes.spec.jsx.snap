// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Downvote component should mount without crashing 1`] = `
ReactWrapper {
  "length": 1,
  Symbol(enzyme.__unrendered__): <Downvotes
    downvotes={1}
    recipeId={1}
    toggleVote={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "downvotes": 1,
      "recipeId": 1,
      "toggleVote": [MockFunction],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": <i
          aria-hidden="true"
          class="fa fa-thumbs-down text-danger click-add"
          id="downvote-button"
        />,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "aria-hidden": "true",
          "className": "fa fa-thumbs-down text-danger click-add",
          "id": "downvote-button",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": Array [
          undefined,
        ],
        "type": "i",
      },
      Object {
        "instance": <span
          class="downvote"
        >
          1
        </span>,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": 1,
          "className": "downvote",
        },
        "ref": null,
        "rendered": Array [
          1,
        ],
        "type": "span",
      },
      " ",
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "downvotes": 1,
        "recipeId": 1,
        "toggleVote": [MockFunction],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": <i
            aria-hidden="true"
            class="fa fa-thumbs-down text-danger click-add"
            id="downvote-button"
          />,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "aria-hidden": "true",
            "className": "fa fa-thumbs-down text-danger click-add",
            "id": "downvote-button",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Array [
            undefined,
          ],
          "type": "i",
        },
        Object {
          "instance": <span
            class="downvote"
          >
            1
          </span>,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": 1,
            "className": "downvote",
          },
          "ref": null,
          "rendered": Array [
            1,
          ],
          "type": "span",
        },
        " ",
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
